components/nonmem/src/metadata.rs

```diff
@@ -0,0 +1,243 @@
+    let model_dir = model_dir.as_ref();
+    for based_on_path in based_on_vec {
+        let full_path = model_dir.join(based_on_path);
+        if !full_path.exists() {
+            bail!(
+                "Based-on model file does not exist: {} (resolved to {})",
+                based_on_path,
+                full_path.display()
+            );
+        }
+    }
+    Ok(())
+}
+
+// helper to take metadata file and get mod/ctl file
+fn resolve_model_path(input: impl AsRef<Path>) -> Result<PathBuf> {
+    let input = input.as_ref();
+    match input.extension().and_then(|e| e.to_str()) {
+        Some("mod") | Some("ctl") => Ok(input.to_path_buf()),
+        _ => {
+            let name = input
+                .file_name()
+                .ok_or_else(|| anyhow!("no filename"))?
+                .to_string_lossy();
+            let base = name
+                .strip_suffix("_metadata.json")
+                .ok_or_else(|| anyhow!("expected '*_metadata.json'"))?;
+            let dir = input.parent().unwrap_or_else(|| Path::new(""));
+
+            let mod_path = dir.join(format!("{base}.mod"));
+            if mod_path.exists() {
+                return Ok(mod_path);
+            }
+
+            let ctl_path = dir.join(format!("{base}.ctl"));
+            if ctl_path.exists() {
+                return Ok(ctl_path);
+            }
+
+            bail!("no .mod or .ctl next to {}", input.to_string_lossy());
+        }
+    }
+}
+
+pub fn update_metadata_file(
+    input: PathBuf,
+    description: Option<String>,
+    tags: Vec<String>,
+    based_on: Vec<String>,
+    overwrite: bool,
+) -> Result<PathBuf> {
+    let model_path = resolve_model_path(&input)?;
+    let (model_name, model_dir) = validate_model_path(&model_path)?;
+    let metadata_path = model_dir.join(format!("{model_name}_metadata.json"));
+
+    let tags_vec = clean_vec(tags);
+    let based_on_vec = clean_vec(based_on);
+
+    validate_based_on(&based_on_vec, &model_dir)?;
+
+    let metadata = if metadata_path.exists() {
+        let m = ModelMetadata::load(&metadata_path)?;
+        m.update(description, tags_vec, based_on_vec, overwrite)
+    } else {
+        // Create new metadata with provided fields
+        let mut m = ModelMetadata::new(based_on_vec, description.unwrap_or(String::new()))?;
+        m.tags = tags_vec;
+        m
+    };
+
+    metadata.save(&model_name, &model_dir)?;
+    Ok(metadata_path)
+}
+
+pub fn clear_metadata_file(
+    model_path: PathBuf,
+    all: bool,
+    clear_based_on: bool,
+    clear_tags: bool,
+) -> Result<PathBuf> {
+    let (model_name, model_dir) = validate_model_path(&model_path)?;
+    let metadata_path = model_dir.join(format!("{model_name}_metadata.json"));
+
+    // Check if metadata file exists
+    if !metadata_path.exists() {
+        bail!(
+            "Metadata file does not exist: {}. Use 'set' or 'append' to create metadata first.",
+            metadata_path.display()
+        );
+    }
```

Comment:
I feel like those checks should be in the CLI directly. Is it because you're planning to expose those methods in hyperion?